proposals/0014-wave-size-range.md

```diff
@@ -0,0 +1,320 @@
+ optimal wave size for a given shader in addition to the range of acceptable
+ sizes.
+This also preserves the ability to force the driver to choose your preferred
+ wave size when multiples are available just as was done with the single value
+ `WaveSize` attribute.
+
+## Detailed Design
+
+### HLSL additions
+
+The existing `WaveSize` attribute gains a second overload:
+
+```HLSL
+[WaveSize(<minWaveSize>, <maxWaveSize>, [prefWaveSize])]
+void main() ...
+```
+
+Where `minWaveSize` is the minimum wave size supported by the shader
+ representing the beginning of the allowed range,
+ `maxWaveSize` is the maximum wave size supported by the shader
+ representing the end of the allowed range,
+ and `prefWaveSize` is the optional preferred wave size representing the size
+ expected to be the most optimal for this shader.
+
+### DXIL Additions
+
+The existing `WaveSize` value is stored as a metadata 32-bit integer constant.
+To store the additional values, an additional metadata tag indicating a tuple of
+ three 32-bit integers representing the given parameters to `WaveSize`.
+In the case where no preferred wave size is specified, the value zero will
+ indicate that nothing was specified.
+
+|         Tag             | Constant |         Value           |
+|-------------------------|----------|-------------------------|
+|kDxilRangedWaveSizeTag   |    23    |MD list: (i32, i32, i32) |
+
+### SPIR-V Additions
+
+To represent the wave size range and preferred value in SPIR-V,
+ a new `ExecutionMode` value would need to be defined for the `OpExecutionMode`
+ instruction to allow specifying `SubgroupSize` with three operands instead of
+ one.
+Like the other formats, the location of the operands would indicate
+ what each value represents, the first two being the min and max wave sizes and
+ the third being the preferred wave size.
+
+### Diagnostic Changes
+
+#### New Errors
+
+These are where new or slightly altered errors are produced:
+
+* If any of the parameters of `WaveSize` are not literal power of two integers
+  between 4 and 128 inclusive.
+* If the minimum wave size value is greater than or equal to the max wave size
+ as respectively specified by the first and second parameters in the `WaveSize`
+ attribute.
+* If the preferred wave size value indicated by the third parameter of
+ `WaveSize` is not in the range specified by the first two parameters.
+* If multiple `WaveSize` attributes are applied to the same entry point
+ with different numbers or values of parameters,
+ the existing error indicating an attribute conflict is produced.
+* If more than three or fewer than one parameter is applied to `WaveSize`.
+* If negative values are provided for any of the `WaveSize` parameters.
+* If float values are provided for any of the `WaveSize` parameters.
+* If non-numerical values are provided for any of the `WaveSize` parameters.
+* If non-literal variables are provided for any of the `WaveSize` parameters.
```

Comment:
Random thought. Do these values really need to be literals or could they be compile-time constant expressions? Is there any reason not to support something like:

```hlsl
static const int Val = 2;
[WaveSize(Val, Val * 4, Val * 2)]
```

I realize that we have some bugs in DXC that relate to evaluation of constant expressions, but we could certainly support this use case in Clang.

Reply:
I agree we should support this, and I believe we do support some variations of constant expressions in similar attribute cases.  This is probably an issue for follow up on what terminology we should use for these attribute values, and figure out what DXC currently supports.  Could you file an issue for later follow up so we can merge the proposal?

Reply:
https://github.com/microsoft/hlsl-specs/issues/161